fairlearn/metrics/_metric_frame.py

```diff
@@ -194,24 +207,181 @@ def __init__(self,
                  y_pred,
                  sensitive_features,
                  control_features: Optional = None,
-                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None):
+                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None,
+                 streaming: bool = False):
         """Read a placeholder comment."""
-        check_consistent_length(y_true, y_pred)
+        if any(arg is None or len(arg) == 0 for arg in
+               (y_true, y_pred, sensitive_features)) and not streaming:
+            raise ValueError(_BAD_INIT_ERROR)
+
+        self._sensitive_features = sensitive_features
+        self._control_features = control_features
+        self._sample_params = sample_params
+        self._streaming = streaming
+        self.metrics = metrics
+
+        self._cf_names = None
+        self._sf_names = None
+
+        if not streaming:
+            # Since we are not streaming, we can compute the metrics already.
+            # This keeps the existing behavior where we compute the metrics in the
+            # constructor.
+            check_consistent_length(y_true, y_pred)
+            y_true_processed = _convert_to_ndarray_and_squeeze(y_true)
+            y_pred_processed = _convert_to_ndarray_and_squeeze(y_pred)
+            self._compute_metric(y_true_processed, y_pred_processed,
+                                 sensitive_features,
+                                 control_features,
+                                 sample_params)
+        else:
+            # If we are streaming, we wait, but initialize accumulators.
+            accumulators = [y_true, y_pred, sensitive_features]
+            # control_features must be None or an empty list.
+            cf_validated = control_features is None or len(control_features) == 0
+            # Same for sample_params
+            sp_validator = sample_params in (None, {})
+            if any(acc != [] for acc in accumulators) or not cf_validated or not sp_validator:
+                raise ValueError(_NON_EMPTY_INIT_ERR)
+            self._y_true = []
+            self._y_pred = []
+            self._overall = None
+            self._by_group = None
+
+    def add_data(self, *, y_true, y_pred, sensitive_features, control_features: Optional = None,
+                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None):
+        """Add data to the MetricFrame.
+
+        Parameters
+        ----------
+        y_true : List, pandas.Series, numpy.ndarray, pandas.DataFrame
```

Comment:
```suggestion
        y_true : array-like
```

?

Reply:
Doesn't sphinx complain about this since it's not technically a "type"?

Reply:
I don't think so: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html#sklearn.decomposition.FastICA.fit_transform

Reply:
This is the guidelines sklearn has: https://scikit-learn.org/stable/developers/contributing.html?highlight=contribute#guidelines-for-writing-documentation